### DISCUSSION TITLE
Deploy Github Pages website and set the environment

### DISCUSSION THREAD

**Discussion by Guillaume-Garrigos** [jupyterlite/jupyterlite#902] [Q&A] [answered]
Hi,

First of all, bravo for this project which looks really great!

I'm starting to play around with it, and one thing I'd like to be able to do is to set the python environment with some preinstalled packages. 

In [this](https://jupyterlite.readthedocs.io/en/latest/quickstart/deploy.html) tutorial, it is said that we can do so by modifying the `requirements.txt `. But actually the situation seems to be more complicated than it appears, or at least it is confusing enough for me.
- For instance `ipywidgets` appearing in `requirements.txt ` does not mean that it is "installed". I barely understand what I am talking about, but when opening a notebook with jupyterlite, `import ipywidgets` returns an error. This is discussed [here](https://github.com/jupyterlite/jupyterlite/discussions/236).
- Another confusing thing is that some packages i've put in `requirements.txt ` get installed, some others don't. For instance `addict` doesn't get installed, but `autograd` does. (Very confusing to me)

I see that I can install things on-the-fly with piplite, and it works well! But the thing is that I would like to avoid doing so. Ideally I wouldn't have to modify all my notebooks to add some code to install missing packages. I'd like my environment to be set once and for all.

So my question is : what am I missing here, and why isn't `requirements.txt ` working the way I expect? Am I doing something wrong? If not, is [Xeus-Python](https://xeus-python-kernel.readthedocs.io/en/latest/) what I want?

Thanks for your patience!

**Comment by jtpio** [comment, marked as answer]
Thanks @Guillaume-Garrigos.

Yes managing dependencies so they are correctly installed in both the kernel and the frontend (UI) can still be a bit challenging at the moment.

There should be some pointers in the documentation already: https://jupyterlite.readthedocs.io/en/latest/howto/configure/simple_extensions.html#the-case-of-jupyter-widgets-and-custom-renderers. The takeover is that dependencies need to be installed in two places, at least for now.

`xeus-python` makes it a bit easier to some extent, by pre-bundling the dependencies at build time so no extra `piplite.install()` is required after starting a kernel. However they still need to be specified as part of `XeusPythonEnv`: https://github.com/jupyterlite/xeus-python-kernel/issues/53

**Reply by Guillaume-Garrigos** [reply]
If you have time, would you mind to tell me where to put this `XeusPythonEnv.packages` in the repository? I couldn't find any details on this in the doc.

**Reply by bollwyvl** [reply]
https://xeus-python-kernel.readthedocs.io/en/latest/configuration.html#pre-installed-packages

**Reply by Guillaume-Garrigos** [reply]
I saw this, but it seems to be isntructions when deploying jupyterlab from the command line? 

What I am doing (and I realize now that I never made it explicit in the conversation) is that I am using a clone of the jupyterlite demo repo. I thought that I could set the environment for xeus by just putting a proper file in the repo.

I guess that I could achieve what I want by changing the Github action [here](https://github.com/jupyterlite/demo/blob/main/.github/workflows/deploy.yml)? I'm going to give it a look.

**Reply by Guillaume-Garrigos** [reply]
Ok so I can indeed change the github action to include the option `jupyter lite build --XeusPythonEnv.packages=list_of_packages` which works quite well.
But I am still trying to figure out how to use "The same can be achieved through a jupyterlite_config.json file" from the documentation. Putting such file at the root of the repo doesn't seem to work.

**Reply by jtpio** [reply]
@Guillaume-Garrigos you can have a look at the following configuration as an example: https://github.com/jtpio/scipy-2022-virtual-poster/blob/ec8ea946c62b4b0ae13d6c68f2255cf52174b127/jupyter_lite_config.json

Placing the file at the top-level should work.

**Reply by Guillaume-Garrigos** [reply]
That is what I originally tried, but it does not work, or at least partially (please tell me if I cant keep the conversation going like this). In view to help you to help me, I am providing a few links below to my repo and Github Page.

More precisely, I want to load the modules `ipympl` and `addict`. The both appear in my `jupyterlite_config.json` file as it can be seen [here](https://github.com/Guillaume-Garrigos/jupyterlite_test/blob/main/jupyterlite_config.json). Just in case they also are in my `requirements.txt` [file](https://github.com/Guillaume-Garrigos/jupyterlite_test/blob/main/requirements.txt).

When reaching the [Lite page](https://guillaume-garrigos.com/jupyterlite_test/lab/index.html), opening `import_test.ipynb`, selecting a XPython kernel, the import of those two modules fail. Am I missing something obvious or is it a current limitation?

One thing I was thinking is that `addict` is available on Pypi (not conda) and that could cause problems. But `ipympl` is on conda-forge so it can't be this?

**Comment by psychemedia** [comment]
Does the repo at https://github.com/jupyterlite/xeus-python-demo also help? It's a template repo for creating a jupyterlite Github Pages site backed by a xeus-python kernel with preinstalled packages.
